Introduce new middleware to track & collect utilization metrics

carlosantoniodasilva opened the pull request:
The middleware will track a counter of inflight requests and push active
requests and whether that represents an active process with a configured
interval. We'll explore that information to be used as an "utilization"
autoscaling metric.

Those metrics go to the same store as the other metrics (queue time,
etc.) so that they're included with the reports to our backend
automatically.

Note that the middleware is only available in judoscale-rails for now,
and is not included by default, it needs to be manually enabled via
config or ENV var.

carlosantoniodasilva committed:
Rename test to match implementation

carlosantoniodasilva committed:
Ensure the request middleware returns the app's response

carlosantoniodasilva committed:
Introduce new middleware to track & collect utilization metrics

The middleware will track a counter of inflight requests and push active
requests and whether that represents an active process with a configured
interval. We'll explore that information to be used as an "utilization"
autoscaling metric.

Those metrics go to the same store as the other metrics (queue time,
etc.) so that they're included with the reports to our backend
automatically.

Note that the middleware is only available in judoscale-rails for now,
and is not included by default, it needs to be manually enabled via
config or ENV var.

carlosantoniodasilva committed:
Give it a larger delta as tests may be slower to run on GH CI

carlosantoniodasilva commented on judoscale-rails/lib/judoscale/rails/railtie.rb:
```diff
@@ -33,6 +34,10 @@ def judoscale_config
+
+        if judoscale_config.utilization_enabled
+          app.middleware.insert_after RequestMiddleware, UtilizationMiddleware, interval: judoscale_config.utilization_interval
+        end
```
There's a potential problem with pushing the middleware conditionally like this, is that by this time Rails hasn't run `config/initializers/*` so anyone trying to set the config to enable/disable this will not work, only via ENV var.

There doesn't seem to be an initializer we can use, as `after_initialize` already has the middleware built / frozen. The only other option I could think of is to always include the initializer, but conditionally run the tracking in there... or hooking into `RequestMiddleware` to do it on a single middleware maybe?

adamlogic replied:
> The only other option I could think of is to always include the initializer, but conditionally run the tracking in there... or hooking into RequestMiddleware to do it on a single middleware maybe?

I think hooking into `RequestMiddleware` would be fine, but keeping the logic in a separate class/module.

carlosantoniodasilva replied:
Another thought came to mind this morning: we can setup our initializer to run specifically after [Rails' `load_config_initializers`](https://github.com/rails/rails/blob/ef88965e8a0c72496c210a5a0a48b85ec9a2ed17/railties/lib/rails/engine.rb#L641-L645), which should load all `config/initializers/*` code, so by the time we get here any Judoscale custom config should be set for us to rely on. This means that setting `config.utilization_enabled = true` would work.

adamlogic replied:
Would that be _instead of_ these Railtie `initializer` blocks? As long as that's supported in all versions of Rails that we support, that sounds great!

carlosantoniodasilva replied:
I set it up as a new initializer to not mess with the existing one (but it'd be okay to change that too I think), basically you just tell it to run `after: :load_config_initializers` which is the Rails one:

https://github.com/judoscale/judoscale-ruby/blob/39bcb4379c4f58fa5698b5cf8b76dbfce2079245/judoscale-rails/lib/judoscale/rails/railtie.rb#L39-L43

That initializer is documented on Rails 6.1+ so we should be okay: https://guides.rubyonrails.org/v6.1/configuring.html#initializers 
![Screenshot 2025-04-17 at 16 36 02](https://github.com/user-attachments/assets/a1e749fd-c936-4aaf-b805-2f2047dc2de7)

adamlogic replied:
Ohhh got it. I thought you were saying we could put _our code_ into `config/initializers` and avoid the Railtie, but I'm with you now—we need those initializers to _run first_. 👍

carlosantoniodasilva commented on judoscale-rails/lib/judoscale/rails/config.rb:
```diff
@@ -3,12 +3,15 @@
         @rake_task_ignore_regex = /assets:|db:/
+
+        @utilization_enabled = ENV["JUDOSCALE_UTILIZATION_ENABLED"] == "true"
+        @utilization_interval = (ENV["JUDOSCALE_UTILIZATION_INTERVAL"] || 1.0).to_f
```
I thought of switching this interval to be in seconds, to match the report interval... also maybe we don't need to be as granular, every ~1 second is enough of a default to track/push metrics?

adamlogic replied:
I'm skeptical that we'd lose too much data with 1 second, but we can play with it.

Totally good with the config in seconds vs milliseconds, though.

nateberkopec replied:
@adamlogic now that this is in GA, can we remove this check?

nateberkopec replied:
(er, the enabled check, not the interval check)

adamlogic replied:
Not quite GA yet... we still need to add this to our Node and Python adapters. We'll remove this check soon.

carlosantoniodasilva commented on judoscale-rails/lib/judoscale/rails/config.rb:
```diff
@@ -3,12 +3,15 @@
   module Rails
     module Config
-      attr_accessor :start_reporter_after_initialize, :rake_task_ignore_regex
+      attr_accessor :start_reporter_after_initialize, :rake_task_ignore_regex, :utilization_enabled, :utilization_interval
```
My first implementation had utilization be a config object so you could do `utilization.enabled` and `utilization.interval`, but I bailed on that to keep it simpler.

carlosantoniodasilva commented on judoscale-ruby/test/request_middleware_test.rb:
```diff
@@ -35,9 +35,11 @@ def call(env)
+          response = middleware.call(env)
+
+          _(response).must_equal app
+          _(app.env).must_equal env
```
Expanding tests a bit to make sure the response is actually returned from the middleware always.

carlosantoniodasilva commented on judoscale-rails/lib/judoscale/rails/utilization_middleware.rb:
```diff
@@ -0,0 +1,70 @@
+
+          Thread.new do
+            # Advise multi-threaded app servers to ignore this thread for the purposes of fork safety warnings.
+            Thread.current.thread_variable_set(:fork_safe, true)
```
Setting this as we're doing with the reporter.

carlosantoniodasilva reviewed (commented):
@adamlogic I'm going to look into the test & lint failure, but I had a couple of questions up for discussion here that I wanted to share initially. LMK what you think.

adamlogic reviewed (commented):
Looking good! Those tests would've taken me an eternity to write. 😅

carlosantoniodasilva committed:
Initialize utilization middleware after app config initializers

This allows users to setup the Judoscale configuration to enable
utilization in `config/initializers`, otherwise we'd run our initializer
check before Rails had a chance to run `config/initializers` code, and
changes there would have no effect. By setting our middleware
initializer to run after `load_config_initializers` from Rails,
Judoscale custom config should be all set for us to rely on here.

carlosantoniodasilva committed:
Remove extraneous nesting with describe

`#call` is the only relevant thing this middleware does, no need to nest
it within a describe block, keep it simple.

carlosantoniodasilva committed:
Make standard happy

carlosantoniodasilva committed:
Stop running utilization tracking tests through the thread

There seems to be some weird thing going on with this test when running
via the utilization tracker thread and the loop/sleep stubs on Ruby 2.7,
it randomly fails as if `Thread#join` didn't do anything / waited for
the thread to run/finish. This test fails on CI under Ruby 2.7 on almost
every run, changing between AR 7.0 and AR 7.1 versions, but always on
Ruby 2.7, and it never failed on other versions of Ruby.

Instead of running through the thread, let's assume the thread is doing
the right thing, and just test that the tracking logic works / pushes
metrics to be reported as we expect.

carlosantoniodasilva commented on judoscale-rails/test/utilization_middleware_test.rb:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    it "tracks utilization metrics for active requests" do
+      tracker.track_current_state
```
I was going through the thread here to test these, but for some reason tests kept failing randomly on Ruby 2.7, and I was kinda able to reproduce locally (albeit still randomly), but could not figure out the root cause, it seemed as if `Thread#join` wasn't being respected, and the code within the thread never ran when the test failed -- but it'd take several successful runs for it to eventually fail once or twice.

So I swapped the test to call the track state method directly on these, and avoid the thread altogether for now.

carlosantoniodasilva reviewed (commented):
Thanks for the review! I had to tweak tests a bit since they were causing trouble on Ruby 2.7, but I got everything green now, good for another 👀 .

adamlogic reviewed (approved):
Looks great! 👍
